Modules/rules.py

```diff
@@ -0,0 +1,132 @@
+import logging
+import re
+from typing import Callable, NamedTuple, Pattern, List, Tuple, Optional
+
+log = logging.getLogger(__name__)
+
+_rules: List["Rule"] = []
+_prefixless_rules: List["Rule"] = []
+
+
+class Rule(NamedTuple):
+    pattern: Pattern
```

Comment:
Are these intentionally class attributes as opposed to instance attributes?

Reply:
This is the 'new' syntax for NamedTuple. It looks at the class annotations and constructs a NamedTuple from it. This is equivalent to `Rule = NamedTuple("Rule", pattern=Pattern, ...)`, however it allows me to define additional methods.